pkg/model/migrator/validator.go

```diff
@@ -0,0 +1,215 @@
+package migrator
+
+import (
+	"bytes"
+	"crypto"
+	"encoding"
+	"fmt"
+
+	"github.com/gohornet/hornet/pkg/whiteflag"
+	"github.com/iotaledger/iota.go/address"
+	"github.com/iotaledger/iota.go/api"
+	"github.com/iotaledger/iota.go/bundle"
+	legacy "github.com/iotaledger/iota.go/consts"
+	"github.com/iotaledger/iota.go/encoding/b1t6"
+	"github.com/iotaledger/iota.go/encoding/t5b1"
+	"github.com/iotaledger/iota.go/merkle"
+	"github.com/iotaledger/iota.go/transaction"
+	"github.com/iotaledger/iota.go/trinary"
+	"github.com/iotaledger/iota.go/v2"
+
+	_ "golang.org/x/crypto/blake2b" // import implementation
+)
+
+var hasher = whiteflag.NewHasher(crypto.BLAKE2b_512)
+
+// Validator takes care of fetching and validating white-flag confirmation data from legacy nodes
+// and wrapping them into receipts.
+type Validator struct {
+	api *api.API
+
+	coordinatorAddress         trinary.Hash
+	coordinatorMerkleTreeDepth int
+}
+
+// NewValidator creates a new Validator.
+func NewValidator(api *api.API, coordinatorAddress trinary.Hash, coordinatorMerkleTreeDepth int) *Validator {
+	return &Validator{
+		api:                        api,
+		coordinatorAddress:         coordinatorAddress,
+		coordinatorMerkleTreeDepth: coordinatorMerkleTreeDepth,
+	}
+}
+
+// QueryMigratedFunds queries the legacy network for the white-flag confirmation data for the given milestone
+// index, verifies the signatures of the milestone and included bundles and then compiles a slice of migrated fund entries.
+func (m *Validator) QueryMigratedFunds(milestoneIndex uint32) ([]*iota.MigratedFundsEntry, error) {
+	confirmation, err := m.api.GetWhiteFlagConfirmation(milestoneIndex)
+	if err != nil {
+		return nil, fmt.Errorf("API call failed: %w", &SoftError{Err: err})
+	}
+
+	included, err := m.validateConfirmation(confirmation, milestoneIndex)
+	if err != nil {
+		return nil, fmt.Errorf("invalid confirmation data: %w", &CriticalError{Err: err})
+	}
+
+	migrated := make([]*iota.MigratedFundsEntry, 0, len(included))
+	for i := range included {
+		output := included[i][0]
+		edAddr, _ := address.ParseMigrationAddress(output.Address)
```

Comment:
I've added a comment but essentially because we're already verifying beforehand that all the bundles are indeed migration bundles and therefore the tail tx is the output tx with the valid migration address.